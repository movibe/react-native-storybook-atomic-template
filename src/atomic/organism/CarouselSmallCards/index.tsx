import React, { useState } from 'react'
import { Dimensions } from 'react-native'
import Carousel, { Pagination } from 'react-native-snap-carousel'

import { Row } from '../../../atomic/atoms/Flex'
import { SmallCard, ISmallCardProps } from '../../../atomic/molecules/SmallCard'

export interface ICarouselSmallCardsProps {
  layout?: 'default' | 'stack' | 'tinder'
  testID?: string
  cards: ISmallCardProps[]
  sliderWidth?: number
  itemWidth?: number
  autoplay?: boolean
  onPress?: () => void
  pagination?: boolean
}

export const CarouselSmallCards: React.FC<ICarouselSmallCardsProps> = ({
  testID = 'CarouselSmallCards',
  layout = 'default',
  cards,
  sliderWidth = Dimensions.get('window').width,
  itemWidth = 100,
  onPress,
  autoplay,
  pagination,
}): JSX.Element => {
  const [Active, setActive] = useState(0)
  const _renderItem = ({ item }) => <SmallCard testID={`${testID}_item`} {...item} onPress={onPress} width={itemWidth} />

  return (
    <>
      <Row testID={testID}>
        <Carousel
          autoplay={autoplay}
          layout={layout}
          data={cards}
          renderItem={_renderItem}
          sliderWidth={sliderWidth}
          itemWidth={itemWidth}
          onSnapToItem={setActive}
        />
      </Row>
      {pagination && (
        <Row justifyContent='center' alignItems='center' style={{ height: 40 }}>
          <Pagination
            dotsLength={cards.length}
            activeDotIndex={Active}
            dotStyle={{
              width: 10,
              height: 10,
              borderRadius: 5,
              backgroundColor: 'rgba(0, 0, 0, 0.92)',
            }}
            inactiveDotStyle={{
              backgroundColor: 'rgba(255, 0, 0, 0.92)',
            }}
            inactiveDotOpacity={0.4}
            inactiveDotScale={0.6}
          />
        </Row>
      )}
    </>
  )
}
