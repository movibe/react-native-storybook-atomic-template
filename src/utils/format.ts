export const formatPoints = (points: string | number): string => {
  const minus = Number(points) < 0

  let n = Number(points).toString()
  let r = ''
  let x = 0

  for (let i = n.length; i > 0; i--) {
    r += n.substr(i - 1, 1) + (x == 2 && i != 1 ? '.' : '')
    x = x == 2 ? 0 : x + 1
  }
  return `${minus ? '-' : ''}${r.split('').reverse().join('')}`
}

export interface IFlightsStops {
  name: string
  iata: string
  duration: string
}
