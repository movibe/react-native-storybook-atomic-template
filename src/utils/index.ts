export * from './faker'
export * from './qualityAssurance'
export * from './format'
export * from './withTheme'
