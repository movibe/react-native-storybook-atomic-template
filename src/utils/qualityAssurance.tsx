import { Platform } from 'react-native'

interface ItestID {
  accessible?: boolean
  testID: string
  accessibilityLabel?: string
}
const testID = (testID: string): ItestID =>
  Platform.OS === 'android'
    ? {
        testID,
        accessible: true,
        accessibilityLabel: testID,
      }
    : {
        testID,
        accessible: false,
      }

type componentType =
  | 'ActivityIndicator'
  | 'View'
  | 'Lottie'
  | 'ScrollView'
  | 'Modal'
  | 'TextInput'
  | 'Image'
  | 'Text'
  | 'ImageBackground'
  | 'TouchableOpacity'
  | 'Radio'
  | 'Checkbox'

export const assignTestId = (componentType: componentType, componentId: string): ItestID => {
  const id = `${componentType}_${componentId}`
  return testID(id)
}
export const mockTestID = (componentType: componentType, componentId: string): ItestID => assignTestId(componentType, componentId)
