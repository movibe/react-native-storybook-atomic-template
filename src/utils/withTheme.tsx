import { boolean, radios } from '@storybook/addon-knobs'
import * as React from 'react'
import styled, { DefaultTheme, ThemeProvider } from 'styled-components/native'

import { Column } from '../atomic/atoms/Flex'
import { themes } from '../tokens'

const options = {
  Light: 'light',
  Dark: 'dark',
}

interface IProps {
  children: any
  padding?: boolean
}

const Background = styled.View`
  background-color: ${({ theme }) => theme.colors.background};
  height: 100vh;
  flex: 1;
`
export const StorybookHoc: React.FC<IProps> = ({ children, padding = true }) => {
  const color = radios('theme_color', options, 'light')
  const _theme: DefaultTheme = color === 'dark' ? themes.dark : themes.light

  return (
    <ThemeProvider theme={_theme}>
      <Background>
        <Column content={boolean('theme_padding', !!padding)} flex={1}>
          {children}
        </Column>
      </Background>
    </ThemeProvider>
  )
}

export const withTheme = (fn: any) => <StorybookHoc>{fn()}</StorybookHoc>
