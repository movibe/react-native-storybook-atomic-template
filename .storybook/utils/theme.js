import { create } from '@storybook/theming/create'

import brandImage from '../../src/assets/images/logo.svg'
import { lightTheme } from '../../src/tokens'
import app from '../../app.json'

export const theme = create({
  base: 'light',
  brandTitle: app.name,
  brandUrl: app.url,
  brandImage,
  colorPrimary: lightTheme.colors.main,
  colorSecondary: lightTheme.colors.main,
  // appBg: lightTheme.colors.background,
})
