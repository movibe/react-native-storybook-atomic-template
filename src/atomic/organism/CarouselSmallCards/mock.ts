import { boolean } from '@storybook/addon-knobs'

import { ISmallCardProps } from '../../../atomic/molecules/SmallCard'
import { faker } from '../../../utils'

export const generateSmallCards = (qtd: number = 5, width: number = 150, height: number = 150): ISmallCardProps[] => {
  let cards: ISmallCardProps[] = []

  for (let index = 0; index < qtd; index++) {
    cards.push({
      image: {
        uri: faker.image.image(),
      },
      gradient: boolean('gradient', false),
      text: faker.commerce.productName(),
      width,
      height,
    })
  }

  return cards
}
