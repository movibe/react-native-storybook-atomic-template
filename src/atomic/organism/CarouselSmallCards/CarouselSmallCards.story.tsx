import { action } from '@storybook/addon-actions'
import { boolean, number } from '@storybook/addon-knobs'
import * as React from 'react'

import { CarouselSmallCards } from '../../../atomic/organism/CarouselSmallCards'
import { generateSmallCards } from './mock'

export default {
  title: 'Organisms/CarouselSmallCards',
}

export const normal = () => (
  <CarouselSmallCards
    autoplay={boolean('autoplay', true)}
    pagination={boolean('pagination', true)}
    cards={generateSmallCards(number('cards', 5))}
    itemWidth={number('itemWidth', 145)}
    onPress={action('onPress')}
  />
)

export const stack = () => (
  <CarouselSmallCards
    layout='stack'
    autoplay={boolean('autoplay', true)}
    pagination={boolean('pagination', true)}
    cards={generateSmallCards(number('cards', 5))}
    itemWidth={number('itemWidth', 145)}
    onPress={action('onPress')}
  />
)
export const tinder = () => (
  <CarouselSmallCards
    layout='tinder'
    autoplay={boolean('autoplay', true)}
    pagination={boolean('pagination', true)}
    cards={generateSmallCards(number('cards', 5))}
    itemWidth={number('itemWidth', 145)}
    onPress={action('onPress')}
  />
)
