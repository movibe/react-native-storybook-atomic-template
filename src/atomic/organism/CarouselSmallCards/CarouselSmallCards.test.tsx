import * as React from 'react'
import { render } from 'react-native-testing-library'

import { CarouselSmallCards, ICarouselSmallCardsProps } from '../../../atomic/organism/CarouselSmallCards'
import { mockTestID, StorybookHoc as wrapper } from '../../../utils'

describe('Organisms/CarouselSmallCards', () => {
  let props: ICarouselSmallCardsProps = {
    testID: 'CarouselSmallCards',
    cards: [
      {
        image: {
          uri: 'teste',
        },
        gradient: false,
        text: 'Oi',
      },
    ],
  }

  const { toJSON, findByTestId } = render(<CarouselSmallCards {...props} />, {
    wrapper,
  })

  const el_container = mockTestID('View', props.testID!).testID

  it(`render component #${el_container}`, async () => {
    const sut = await findByTestId(el_container)
    expect(sut).toBeTruthy()
  })

  it('should render snapshot', () => {
    expect(toJSON()).toMatchSnapshot()
  })
})
