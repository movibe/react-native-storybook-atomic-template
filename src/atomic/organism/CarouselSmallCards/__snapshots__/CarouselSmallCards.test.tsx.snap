// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/CarouselSmallCards should render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "100vh",
      },
    ]
  }
>
  <View
    accessible={false}
    style={
      Array [
        undefined,
        Object {
          "alignItems": undefined,
          "borderColor": undefined,
          "borderWidth": 0,
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": undefined,
          "padding": 10,
        },
      ]
    }
    testID="View_View"
  >
    <View
      accessible={false}
      style={
        Array [
          Object {
            "alignItems": undefined,
            "borderColor": undefined,
            "borderWidth": 0,
            "flex": undefined,
            "flexDirection": "row",
            "justifyContent": undefined,
            "padding": undefined,
          },
          undefined,
        ]
      }
      testID="View_CarouselSmallCards"
    >
      <RCTScrollView
        activeAnimationOptions={null}
        activeAnimationType="timing"
        activeSlideAlignment="center"
        activeSlideOffset={20}
        apparitionDelay={0}
        automaticallyAdjustContentInsets={false}
        autoplay={false}
        autoplayDelay={1000}
        autoplayInterval={3000}
        callbackOffsetMargin={5}
        containerCustomStyle={Object {}}
        contentContainerCustomStyle={Object {}}
        contentContainerStyle={
          Array [
            Object {
              "paddingLeft": 325,
              "paddingRight": 325,
            },
            Object {},
          ]
        }
        data={
          Array [
            Object {
              "gradient": false,
              "image": Object {
                "uri": "teste",
              },
              "text": "Oi",
            },
          ]
        }
        decelerationRate="fast"
        directionalLockEnabled={true}
        disableVirtualization={false}
        enableMomentum={false}
        enableSnap={true}
        firstItem={0}
        getItem={[Function]}
        getItemCount={[Function]}
        hasParallaxImages={false}
        horizontal={true}
        inactiveSlideOpacity={0.7}
        inactiveSlideScale={0.9}
        inactiveSlideShift={0}
        initialNumToRender={13}
        invertStickyHeaders={false}
        inverted={false}
        itemWidth={100}
        keyExtractor={[Function]}
        layout="default"
        lockScrollTimeoutDuration={1000}
        lockScrollWhileSnapping={false}
        loop={false}
        loopClonesPerSide={3}
        maxToRenderPerBatch={27}
        numColumns={1}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onResponderRelease={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        onSnapToItem={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        overScrollMode="never"
        pinchGestureEnabled={false}
        removeClippedSubviews={true}
        renderItem={[Function]}
        scrollEnabled={true}
        scrollEventThrottle={1}
        scrollsToTop={false}
        shouldOptimizeUpdates={true}
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
        slideStyle={Object {}}
        sliderWidth={750}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "flexDirection": "row",
            "width": 750,
          }
        }
        swipeThreshold={20}
        updateCellsBatchingPeriod={50}
        useScrollView={false}
        vertical={false}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={27}
      >
        <View>
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "opacity": 1,
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 100,
                }
              }
            >
              <View
                accessible={false}
                focusable={false}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "borderRadius": 10,
                    "height": 126,
                    "opacity": 1,
                    "overflow": "hidden",
                    "shadowColor": "rgba(0,0,0,0.2)",
                    "shadowOffset": Object {
                      "height": 5,
                      "width": 0,
                    },
                    "shadowOpacity": 1,
                    "shadowRadius": 15,
                    "width": 100,
                  }
                }
                testID="TouchableOpacity_CarouselSmallCards_item"
              >
                <View
                  accessibilityIgnoresInvertColors={true}
                  style={
                    Array [
                      Object {
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "flex-end",
                        "overflow": "hidden",
                      },
                    ]
                  }
                >
                  <Image
                    source={
                      Object {
                        "uri": "teste",
                      }
                    }
                    style={
                      Array [
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        },
                        Object {
                          "height": undefined,
                          "width": undefined,
                        },
                        undefined,
                      ]
                    }
                  />
                  <Text
                    accessible={false}
                    style={
                      Array [
                        Object {
                          "color": "#444",
                          "fontSize": 12,
                          "textTransform": "none",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 14,
                          "fontWeight": "bold",
                          "marginBottom": 10,
                          "marginLeft": 10,
                          "marginRight": 10,
                          "marginTop": 10,
                        },
                      ]
                    }
                    testID="Text_CarouselSmallCards_item"
                  >
                    Oi
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
